sycl/doc/extensions/experimental/sycl_ext_oneapi_copy_optimize.asciidoc

```diff
@@ -0,0 +1,138 @@
+access to features and gather community feedback. Interfaces defined in this
+specification are implemented in DPC\++ but they are not finalized and may
+change incompatibly in future versions of DPC++ without prior notice.
+Shipping software products should not rely on APIs defined in this
+specification.
+
+
+== Overview
+
+SYCL provides explicit copy APIs that copy memory between a device and the host.
+Some backends can optimize these copy operations if the backend knows in
+advance that the host memory will be used for such a copy operation.
+If a SYCL application knows that it will frequently copy to or from the same
+host memory, it can use the APIs in this extension to inform the backend
+about this, which can result in a performance benefit.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine
+the macro `SYCL_EXT_ONEAPI_COPY_OPTIMIZE` to one of the values defined
+in the table below.  Applications can test for the existence of this macro
+to determine if the implementation supports this feature, or applications
+can test the macro's value to determine which of the extension's features
+the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== API of the extension
+
+This extension adds the following free functions:
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+
+void prepare_for_device_copy(void *Ptr, size_t NumBytes,
+                             const context &Context);
+void prepare_for_device_copy(void *Ptr, size_t NumBytes,
+                             const queue &Queue);
+
+void release_from_device_copy(void *Ptr, const context &Context);
+void release_from_device_copy(void *Ptr, const queue &Queue);
+
+} // namespace sycl::ext::oneapi::experimental
+```
+
+Table 1. Functions added by this extension.
+|===
+| Member Function | Description
+
+| `prepare_for_device_copy(void *Ptr, size_t NumBytes, const context &Context)`.
+| Informs the implementation that the host memory range starting at `Ptr` and
+extending for `NumBytes` bytes may be used as either the source or destination
+of an explicit copy operation to a queue using context Context. Calling this
+function may allow the implementation to accelerate these copy operations.
+Generally, it is only useful to call this function if the host memory range will
+be used for several subsequent copy operations; it is not generally useful to
+call it before each individual copy operation.
+
+The behavior is undefined when multiple calls to `prepare_for_device_copy`
+specify memory ranges that overlap, even when using different
+SYCL contexts. No error message is issued.
+
+| `prepare_for_device_copy(void *Ptr, size_t NumBytes,
+                             const queue &Queue)`
```

Comment:
The format of this table cell looks terrible in the HTML render.  Can you use a code block like this to improve the line breaks?

> ````
> ```
> void prepare_for_device_copy(
>   void *Ptr, size_t NumBytes,
>   const queue &Queue)
> ```
> ````

I'm not sure if that works inside a table, so check what the HTML render looks like.  If it works, use it for the other table entries too because their line breaks could also be improved.

Reply:
See this commit for how you can fix the formatting:

https://github.com/intel/llvm/commit/e777249f5f426ddfc5c46d49091ce8803d55e143

Reply:
Thanks for the formatting tip. I struggled with it. Done now.